### ISSUE TITLE
Improve MCMC handling

### ISSUE THREAD

**Issue by st--** [GPflow/GPflow#655]
MCMC in GPflow has a couple of snags that could be much improved:

- you no longer get an acceptance rate, you can get a proxy by calculating len(set(logprobs))/len(logprobs), but this only works in hindsight, and makes it difficult to tune parameters like epsilon & to watch what's going on (e.g. whether acceptance rate is going up or down)

- it would be great to have some way of callbacks for MCMC, too

- --> maybe have an outer (python) loop around inner calls to the tensorflowed HMC for, say, sets of 100 iterations (and then output acceptance rate etc.)?

- model.assign() is a) slow and b) becomes slower, which makes it really impractical to actually evaluate anything as a function of your samples!

When assigning a row of the returned DataFrame to the model, you have to remember to `drop('logprobs', axis=1)`... might be good to add that to the documentation/notebook example.

Also, related to training ... it would be so helpful to have a way to interrupt an optimizer by pressing Ctrl+C and getting the latest state back ...

**Comment by st--** [comment]
And when manually iterating over calls to HMC().sample() (e.g. to print out acceptance rate in between), the graph gets bigger & calls become slower and slower.

**Comment by st--** [comment]
Also see #555

**Comment by hughsalimbeni** [comment]
Absolutely agree with all these points. One thing that would be nice is to have a callback like gpflow_monitor where you can evaluate what you're interested in (e.g. the posterior at a few test points) without having to go back and loop over samples. With big models I've found that storing all the samples gets tricky (e.g. trying to take 1e4 posterior draws from a model with 1e5 params is problematic). Perhaps there should be a `forget_samples` option for these situations.

**Comment by st--** [comment]
This is all moot now that we're based on top of tensorflow 2 and piggyback on tensorflow_probability's MCMC routines.
